Support beacon proxies for Hardhat

ericglau opened the pull request:
Part of https://github.com/OpenZeppelin/openzeppelin-upgrades/issues/342

Adds the following APIs to the Hardhat plugin (see [doc](https://github.com/ericglau/openzeppelin-upgrades/blob/beacon/docs/modules/ROOT/pages/api-hardhat-upgrades.adoc)):
`deployBeacon`
`upgradeBeacon`
`deployBeaconProxy`
`loadProxy` (supports UUPS/Transparent/Beacon proxies)

Modifies the following API:
`prepareUpgrade` (accepts a UUPS/Transparent/Beacon proxy address or a Beacon address)

frangio commented on packages/plugin-hardhat/src/utils/deploy-impl.ts:
```diff
@@ -14,53 +14,152 @@ import {
   opts: Options,
-  proxyAddress?: string,
-): Promise<DeployedImpl> {
+  isBeacon: boolean,
+  proxyOrBeaconAddress?: string,
```
I realize now that we can do something simpler and more focused in `deployImpl`, and I think we should.

It seems the only thing that `isBeacon` affects is how we retrieve the implementation address. So my proposal is to move that logic outside of here, and instead of receiving the address of a proxy or a beacon as an argument, receive the implementation address directly. Doing that, `deployImpl` should be a lot simpler. The `kind` return argument can be ignored for beacon proxies.

Let me know what you think!

ericglau replied:
I agree.  This would allow `deployImpl` to just focus on the core logic of validation and deployment.  The scenario-specific parts could perhaps be moved to the `deployProxyImpl` and `deployBeaconImpl` functions above.

frangio commented on packages/plugin-hardhat/src/utils/impl-address.ts:
```diff
@@ -0,0 +1,79 @@
+  hre: HardhatRuntimeEnvironment,
+  implAddress: string,
+): Promise<utils.Interface | undefined> {
+  if (implAddress === undefined) {
```
This shouldn't happen according to the type of the argument `implAddress: string`.

ericglau replied:
Thanks, missed that while refactoring.

frangio commented on packages/plugin-hardhat/src/utils/impl-address.ts:
```diff
@@ -0,0 +1,79 @@
+        beaconAddress,
+      );
+    } catch (e: any) {
+      // error expected if the address was not a beacon proxy
```
Same as a previous comment, we shouldn't ignore the error if it's a network error or something else.

frangio commented on packages/plugin-hardhat/src/load-proxy.ts:
```diff
@@ -0,0 +1,39 @@
+
+    const contractInterface = await getInterfaceFromManifest(hre, implAddress);
+    if (contractInterface === undefined) {
+      throw new Error(
```
As mentioned previously `UpgradesError` can create a nicer error message here. The error message right now is super long, will wrap, etc., I generally try to make errors look nicer in the console, because I feel otherwise people may ignore them or just be annoyed.

frangio commented on packages/plugin-hardhat/src/deploy-proxy.ts:
```diff
@@ -71,25 +78,29 @@ export function makeDeployProxy(hre: HardhatRuntimeEnvironment): DeployFunction
-    if (initializer === false) {
-      return '0x';
-    }
+export function getInitializerData(
```
Can we move this to a file in utils, now that it's also used by `deploy-beacon-proxy`?

frangio commented on packages/plugin-hardhat/src/utils/impl-address.ts:
```diff
@@ -0,0 +1,79 @@
+ */
+export async function getImplementationAddressFromBeacon(
+  hre: HardhatRuntimeEnvironment,
+  signer: ethers.Signer | undefined,
```
We shouldn't need a signer for this function. It's an optional argument in `getIBeaconFactory`.

frangio reviewed (commented):
Great work!

ericglau commented on packages/plugin-hardhat/src/deploy-beacon.ts:
```diff
@@ -0,0 +1,33 @@
+    const { impl } = await deployBeaconImpl(hre, ImplFactory, opts);
+
+    const UpgradeableBeaconFactory = await getUpgradeableBeaconFactory(hre, ImplFactory.signer);
+    const abi = ImplFactory.interface.format(FormatTypes.json);
```
Agreed. This can be changed to `FormatTypes.minimal` to get a [minimal array](https://docs.ethers.io/v5/api/utils/abi/interface/#Interface--formatting).

ericglau replied:
Addressed in https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/456/commits/d4f23ea45e988919def4a546619f9a5d4635d635

ericglau commented on packages/core/src/manifest.ts:
```diff
@@ -25,14 +26,19 @@ export interface ImplDeployment extends Deployment {
+}
+
+export interface BeaconDeployment extends Deployment {
+  abi: string | string[];
```
Addressed in https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/456/commits/d4f23ea45e988919def4a546619f9a5d4635d635

ericglau commented on packages/plugin-hardhat/src/deploy-beacon-proxy.ts:
```diff
@@ -0,0 +1,91 @@
+    const implAddress = await getImplementationAddressFromBeacon(hre, getSigner(ImplFactoryOrSigner), beaconAddress);
+    let contractInterface = await getInterfaceFromManifest(hre, implAddress);
+
+    if (contractInterface === undefined) {
+      if (ImplFactoryOrSigner instanceof ContractFactory) {
```
Here we use the user provided beacon address to look up the implementation address, then get the interface that we previously stored in the manifest for that implementation address.  Is there any possibility that the manifest is incorrect, aside from corruption?  Considering the implementation itself is immutable.

Although since the interface is used to create the initializer call data, and if both are provided by the user, I agree it makes sense to use both together.  Then if there is a mismatch during deployment, there would be an error anyways.

ericglau commented on packages/plugin-hardhat/src/utils/deploy-impl.ts:
```diff
@@ -14,53 +14,152 @@ import {
+  if (isBeacon) {
+    if (proxyOrBeaconAddress !== undefined) {
+      if (await isTransparentOrUUPSProxy(proxyOrBeaconAddress)) {
+        throw new Error(
```
Agreed. Having a separate user action section does make errors easier to understand.

ericglau commented on packages/core/src/validate/query.ts:
```diff
@@ -144,13 +146,32 @@ export function isUpgradeSafe(data: ValidationData, version: Version): boolean {
 }
 
-export function inferProxyKind(data: ValidationData, version: Version): ProxyDeployment['kind'] {
+export async function inferProxyKind(
```
This seems reasonable.  It isn't really needed in `deployProxy`, and the main purpose was to give a helpful error when attempting to upgrade a beacon proxy through `upgradeProxy` or `prepareUpgrade`.  But those functions could just call `isBeaconProxy()` directly.

Another reason for this original approach though was so the Truffle plugin could also detect previously deployed beacon proxies when the user is trying to upgrade them (and then give an error about mismatching `kind`), rather than infer the wrong `kind`.  Although of course this PR does not cover Truffle and perhaps inferring the wrong `kind` is not much of a problem if it fails gracefully otherwise.

frangio commented on packages/plugin-hardhat/src/deploy-beacon-proxy.ts:
```diff
@@ -46,18 +35,27 @@ export function makeDeployBeaconProxy(hre: HardhatRuntimeEnvironment): DeployBea
+      contractInterface = await getInterfaceFromManifest(hre, implAddress);
+      if (contractInterface === undefined) {
+        throw new UpgradesError(
+          `Beacon's current implementation at address ${implAddress} was not found in the network manifest.`,
```
```suggestion
          `Beacon's current implementation at ${implAddress} is unknown`,
```

frangio commented on packages/plugin-hardhat/src/utils/deploy-impl.ts:
```diff
@@ -1,66 +1,154 @@
+    return await beaconContract.implementation();
+  }
+
+  async function assertNotProxy(address: string) {
```
These helper functions will be useful once when we build the Truffle plugin. I would suggest trying to have them in core. But I think this can be done in the separate PR for Truffle.

ericglau replied:
This function is mainly implementing custom error messages which do not seem to belong in core.  The helper methods used here are part of core.

frangio commented on packages/plugin-hardhat/src/utils/deploy-impl.ts:
```diff
@@ -1,66 +1,154 @@
+  }
+  return deployImpl(deployData, ImplFactory, opts, currentImplAddress);
+
+  async function getBeaconImplementationAddress(beaconAddress: string): Promise<string> {
```
This one would also be a good fit for core, but not in the way it's written now. We would need to change it so it doesn't use Ethers.js. We can just manually craft the `implementation()` function call and use it with `eth_call` through JSON RPC.

frangio reviewed (approved):
Some small suggestions but otherwise good to merge! Thanks.

ericglau commented on packages/plugin-hardhat/src/prepare-upgrade.ts:
```diff
@@ -1,18 +1,28 @@
+    } else if (await isBeaconProxy(provider, proxyOrBeaconAddress)) {
+      const beaconAddress = await getBeaconAddress(provider, proxyOrBeaconAddress);
+      deployedImpl = await deployBeaconImpl(hre, ImplFactory, opts, beaconAddress);
+    } else {
```
Added a check in https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/456/commits/54297730341a409fd926abc8c13be51db9caa12f.

However, if the address has an `implementation()` function that returns something other than an address, the returned result may still look like an address according to `ethers.utils.isAddress()`.  In this case, the plugin would just return that the "address" is not found in the manifest.

This check is not necessary in `defender.proposeUpgrade` at this time, since https://github.com/ericglau/openzeppelin-upgrades/blob/beacon/packages/plugin-defender-hardhat/src/propose-upgrade.ts#L42 would give an error if the address is not a Transparent or UUPS proxy.  In order to add this check, we may need to do some refactoring similar to the suggestion in https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/456#discussion_r768174558

ericglau replied:
Refactored and added this check in `defender.proposeUpgrade()`

ericglau commented:
@frangio Please re-review since a lot of content has been refactored into core.  For your convenience, [here is a diff](https://github.com/ericglau/openzeppelin-upgrades/compare/beacon-already-reviewed...ericglau:beacon?expand=1) with the changes since the last review.

frangio commented on packages/core/src/beacon.ts:
```diff
@@ -0,0 +1,20 @@
+ */
+export async function isBeacon(provider: EthereumProvider, beaconAddress: string) {
+  try {
+    return await getImplementationAddressFromBeacon(provider, beaconAddress);
```
This is just returning the string, some comparison to get a boolean is missing I believe.

frangio reviewed (approved):
LGTM! Feel free to merge after this one fix.
